src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -2,44 +2,103 @@
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
+        private const string defaultExtension = ".tmp";
+
+        /// <summary>
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional.
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension { get; set; } = defaultExtension;
+
         /// <summary>
-        /// returns a TemporaryFile
+        /// Returns a TemporaryFile.
         /// </summary>
         protected override void EndProcessing()
         {
-            string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
+            // Check for invalid characters in extension
+            if (Extension.IndexOfAny(Path.GetInvalidFileNameChars()) != -1)
+            {
+                WriteError(
+                    new ErrorRecord(
+                        new ArgumentException(
+                                StringUtil.Format(UtilityCommonStrings.InvalidCharacterInParameter,
+                                    nameof(Extension), Extension)),
+                        "NewTemporaryInvalidArgument",
+                        ErrorCategory.InvalidArgument,
+                        Extension));
+                return;
+            }
+
+            string temporaryFilePath = null;
+            string tempPath = Path.GetTempPath();
+
             if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    filePath = Path.GetTempFileName();
+                    try
+                    {
+                        temporaryFilePath = Path.GetTempFileName(); // this already creates the temporary file
+                        if (Extension != defaultExtension)
+                        {
+                            // rename file
+                            var temporaryFileWithCustomExtension = Path.ChangeExtension(temporaryFilePath, Extension);
+                            File.Move(temporaryFilePath, temporaryFileWithCustomExtension);
+                            temporaryFilePath = temporaryFileWithCustomExtension;
+                        }
+                        creationOfFileSuccessful = true;
+                        WriteDebug($"Created temporary file {temporaryFilePath}.");
+                    }
+                    catch (IOException) // file already exists -> retry
+                    {
+                        attempts++;
+                        if (temporaryFilePath != null)
+                        {
+                            File.Delete(temporaryFilePath);
+                        }
```

Comment:
What if this throws? We won't retry.

Reply:
Should we then simply try File.Delete with an empty catch block instead?

Reply:
If the retry count is 10, that seems fine. If it was a huge number, then I'd say probably not.  A short sleep in the catch might not hurt either, like 10ms or something.

Reply:
Would it make sense to write a warning if we're not able to delete it?

Reply:
At least a warning seems reasonable. But I am slightly more leading to actually throwing a (terminating) error because if deleting the temporary file that just got created failed then I suspect that in most cases there is rather a fundamental problem with permissions. WDYT?